source/canvasrecorder.ts

```diff
@@ -0,0 +1,188 @@
+import { assert } from './auxiliaries';
+
+/* spellchecker: enable */
+
+
+declare let MediaRecorder: any;
+
+type CanvasCaptureMediaStream = any;
+type MediaRecorder = any;
+type ImagesAvailableCallback = (images: Array<Blob>) => void;
+
+/**
+ * Class to record the contents of a Canvas.
+ * Some time after stopping the CanvasRecorder the recorded images will be available as Array of Blobs. The images can
+ * be retrieved by setting the onImagesAvailable callback.
+ * Exemplary usage:
+ * ```
+ * const recorder = new CanvasRecorder(canvas);
+ * const images = new Array<Blob>();
+ * recorder.onImagesAvailable = (data: Array<Blob>) => images.push(data);
+ * recorder.start(60);
+ * ...
+ * recorder.stop();
+ *
+ * ```
+ */
+export class CanvasRecorder {
+    protected _canvas: Canvas;
+
+    protected _stream: CanvasCaptureMediaStream;
+    protected _recorder: MediaRecorder;
+    protected _state: CanvasRecorder.State;
+
+    protected _images: Array<Blob>;
+    protected _onImagesAvailable: ImagesAvailableCallback | undefined;
+
+    /**
+     * Checks whether recording the canvas is supported.
+     */
+    static isSupported(): boolean {
+        return (window as any).MediaRecorder !== undefined;
+    }
+
+    /**
+     * Checks whether the given MIME type is supported.
+     * @param type - MIME type to check.
+     */
+    static isMIMETypeSupported(type: string): boolean {
+        return MediaRecorder.isTypeSupported(type);
+    }
+
+    /**
+     * Creates a CanvasRecorder. Throws, if it is not supported on the used platform.
+     * @param canvas - The canvas to record.
+     */
+    constructor(canvas: Canvas) {
+        assert(CanvasRecorder.isSupported(), 'Recording the canvas is not supported.');
+
+        this._canvas = canvas;
+        this._state = CanvasRecorder.State.INACTIVE;
+        this._images = new Array<Blob>();
+    }
+
+    /**
+     * Starts recording the the canvas.
+     * If the given fps is 0 it won't automatically record. Instead {@link frame} has to be called every time a new
+     * frame should get recorded.
+     * Must not be called with negative fps, while already recording or with an unsupported MIME type.
+     * @param fps - Maximum fps to record in.
+     * @param mimeType - The MIME video type.
+     */
+    start(fps: number, mimeType = ''): void {
+        assert(fps >= 0, 'FPS has to be positive');
+        assert(this._state === CanvasRecorder.State.INACTIVE, 'Recorder has to be inactive.');
+        assert(CanvasRecorder.isMIMETypeSupported(mimeType), `MIME type: ${mimeType} is not supported.`);
+
+        this._stream = (this._canvas.element as any).captureStream(fps);
+        this._recorder = new MediaRecorder(this._stream, { mimeType });
+        this._images.length = 0;
+
+        this._recorder.ondataavailable = (event: any) => this._images.push(event.data);
+        this._recorder.onstop = () => { if (this._onImagesAvailable) this._onImagesAvailable(this._images); }
```

Comment:
According to https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining it should be possible to use optional chaining like this:
this._recorder.onstop = () => this._onImagesAvailable?.(this._images);

I don't know why but the build fails using optional chaining. I think it worked when I created the PR though. Anyone got an idea what changed? Or if I just misremember it?